### PR TITLE
Make STARTS compatible with Java 9 - 15

### PR DESCRIPTION
**Pull request by jtx1999**
- Removed the dependency of `tools.jar`, which was deprecated since Java 9
 - Instead, load `jdeps` and `com.sun.tools.attach.VirtualMachine` dynamically via reflections.
 - Add Github actions CI for multiple Java versions.
 - Upgrade Jacoco versions, so we won't get `NoSuchFieldException: $jacocoAccess`
 - Upgrade Maven compiler source and target for compatibility with higher Java versions.
 - Add config `-Djdk.attach.allowAttachSelf=true`, which became default to false since Java 9.
 - Use `<code>` tag rather than `<tt>` as `<tt>` is deprecated
   - was getting errors like `/Users/tianxing/Desktop/starts/starts-core/src/main/java/edu/illinois/starts/asm/MethodVisitor.java:35: error: tag not supported in the generated HTML version: tt` on Java 11.

### CODE REVIEW

File: starts-core/pom.xml [hunk 1]
```diff
@@ -37,6 +37,14 @@
           </archive>
         </configuration>
       </plugin>
+        <plugin>
```
Comment (owolabileg) [inline starts-core/pom.xml:40]: Add a comment hear on why this is needed and what would happen without this change. Also, do we need to update the documentation.

File: starts-plugin/pom.xml [hunk 2]
```diff
@@ -30,6 +30,14 @@
 
   <build>
     <plugins>
+    <plugin>
```
Comment (owolabileg) [inline starts-plugin/pom.xml:33]: Add a comment hear on why this is needed and what would happen without this change. Also, do we need to update the documentation?

File: starts-core/src/main/java/edu/illinois/starts/asm/AnnotationVisitor.java [hunk 3]
```diff
@@ -32,8 +32,8 @@
 
 /**
  * A visitor to visit a Java annotation. The methods of this class must be
- * called in the following order: ( <tt>visit</tt> | <tt>visitEnum</tt> |
- * <tt>visitAnnotation</tt> | <tt>visitArray</tt> )* <tt>visitEnd</tt>.
+ * called in the following order: ( <code>visit</code> | <code>visitEnum</code> |
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/asm/AnnotationVisitor.java:35]: Next time, please do not mix commits that change format, with that change code. Also, please modify the commit message to say what happens without this changes.

File: .mvn/jvm.config [hunk 4]
```diff
@@ -0,0 +1 @@
+-Djdk.attach.allowAttachSelf=true
```
Comment (owolabileg) [inline .mvn/jvm.config:1]: Add a comment about why this is better than the previous code, and error that you saw without this change.

File: starts-plugin/pom.xml [hunk 5]
```diff
@@ -33,9 +33,6 @@
      <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-surefire-plugin</artifactId>
-       <configuration>
```
Comment (jtx1999) [inline starts-plugin/pom.xml]: Put the errors seen.

File: .mvn/jvm.config (hunk 4 above)
Comment (owolabileg) [reply .mvn/jvm.config:1]: Also, where is your test to show that this change does not break working functionality?

File: .github/workflows/maven.yml [hunk 6]
```diff
@@ -26,4 +26,4 @@ jobs:
       with:
         java-version: ${{ matrix.java-version }}
     - name: Build with Maven
-      run: mvn -B package --file pom.xml
```
Comment (owolabileg) [inline .github/workflows/maven.yml]: why is this needed? please update commit message.

Comment (owolabileg) [reply .github/workflows/maven.yml]: Does this remove the need for travis? If so, explain.

File: pom.xml [hunk 7]
```diff
@@ -78,7 +78,7 @@
 
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <jacoco.version>0.7.8</jacoco.version>
```
Comment (owolabileg) [inline pom.xml:81]: Can you make GitHub actions to display test coverage of STARTS? Please do that in a new PR.

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java [hunk 8]
```diff
@@ -29,6 +30,10 @@
  */
 public class RTSUtil implements StartsConstants {
     private static final Logger LOGGER = Logger.getGlobal();
+    /** Name of tools.jar in JDK */
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:33]: Please use the right kind of comments

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java [hunk 9]
```diff
@@ -76,11 +81,56 @@ public static void computeAndSaveNewCheckSums(String artifactsDir,
     public static Map<String, Set<String>> runJdeps(List<String> args) {
         StringWriter output = new StringWriter();
         LOGGER.log(Level.FINE, "JDEPS ARGS:" + args);
-        Main.run(args.toArray(new String[0]), new PrintWriter(output));
+
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:84]: Please, add a comment to show that you copied from Ekstazi.

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java [hunk 10]
```diff
@@ -76,11 +81,56 @@ public static void computeAndSaveNewCheckSums(String artifactsDir,
...
+                        .invoke(null, args.toArray(new String[0]), new PrintWriter(output));
+            }
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
         // jdeps can return an empty output when run on .jar files with no .class files
         return output.getBuffer().length() != 0 ? getDepsFromJdepsOutput(output) : new HashMap<String, Set<String>>();
     }
 
+    private static File findToolsJar() {
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:108]: This already exists in AgentLoader. Make it public in AgentLoader and reuse here instead of having duplicate code.

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java [hunk 11]
```diff
@@ -76,11 +81,56 @@ public static void computeAndSaveNewCheckSums(String artifactsDir,
...
+                        .invoke(jdeps, new PrintWriter(output), new PrintWriter(output), args.toArray(new String[0]));
+            } else {
+                // Java 8, load tools.jar
+                URLClassLoader loader = new URLClassLoader(new URL[] { toolsJarFile.toURI().toURL() },
+                        ClassLoader.getSystemClassLoader());
+                Class<?> jdepsMain = loader.loadClass("com.sun.tools.jdeps.Main");
+                jdepsMain.getMethod("run", String[].class, PrintWriter.class)
+                        .invoke(null, args.toArray(new String[0]), new PrintWriter(output));
+            }
+        } catch (Exception ex) {
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:101]: Catch the specific exception instead

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java (hunk 9 above)
Comment (owolabileg) [reply starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:84]: Consider moving this lines 85 to 103 to a new method in AgentLoader, so that all the loading of classes happens there. You should no longer need to make findToolsJar public, and you should only have two lines in this runJDeps method. Please do these in a new PR.

File: starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java [hunk 12]
```diff
@@ -25,26 +29,81 @@ public static boolean loadDynamicAgent() {
             URL agentJarURLConnection = AbstractMojoInterceptor.extractJarURL(agentJarURL);
             return loadAgent(agentJarURLConnection);
         } catch (Exception ex) {
```
Comment (jtx1999) [inline starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:31]: Specify exception

File: starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java [hunk 13]
```diff
@@ -25,26 +29,81 @@ public static boolean loadDynamicAgent() {
...
+            ex.printStackTrace();
             return false;
         }
     }
 
     public static boolean loadAgent(URL aju) throws Exception {
-        attachAgent(aju);
+        URL toolsJarFile = findToolsJar();
+
+        Class<?> vc = loadVirtualMachine(new URL[]{toolsJarFile});
```
Comment (jtx1999) [inline starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:40]: rename to vm

Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:40]: vc --> vm

File: starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java [hunk 14]
```diff
@@ -25,26 +29,81 @@ public static boolean loadDynamicAgent() {
...
+        if (vc == null) {
+            return false;
+        }
+
+        attachAgent(vc, aju);
         return true;
     }
 
-    private static void attachAgent(URL aju) throws Exception {
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java]: Add that we copied from Ekstazi and specify the SHA (commit ID) of Ekstazi that you copied from.

File: starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java [hunk 15]
```diff
@@ -25,26 +29,81 @@ public static boolean loadDynamicAgent() {
...
+            tjf = new File(javaHomeFile, "Classes" + File.separator + CLASSES_JAR_NAME);
+        }
+
+        return tjf.toURI().toURL();
+    }
+
+    private static boolean isMac() {
+        return System.getProperty("os.name").toLowerCase().indexOf("mac") >= 0;
+    }
+}
```
Comment (owolabileg) [inline starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java:109]: add a new line at the end

Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java]: Done

File: starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java (hunk 14 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/maven/AgentLoader.java]: Done

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java (hunk 9 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java:84]: > Please, add a comment to show that you copied from Ekstazi.

Done

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java (hunk 8 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java]: Done

File: .mvn/jvm.config (hunk 4 above)
Comment (jtx1999) [reply .mvn/jvm.config:1]: The previous code is added by me as well. The previous code breaks integration tests on Java 12+: https://github.com/jtx1999/starts/runs/2214242432?check_suite_focus=true

File: starts-core/src/main/java/edu/illinois/starts/asm/AnnotationVisitor.java (hunk 3 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/asm/AnnotationVisitor.java:35]: Done

File: .mvn/jvm.config (hunk 4 above)
Comment (jtx1999) [reply .mvn/jvm.config:1]: The integration tests show that it's not beaking working functionality.

File: .github/workflows/maven.yml (hunk 6 above)
Comment (jtx1999) [reply .github/workflows/maven.yml]: Yes, this removes the need for travis. Since it does the same thing as travis, plus building on multiple java versions.

File: pom.xml (hunk 7 above)
Comment (jtx1999) [reply pom.xml:81]: Done in https://github.com/jtx1999/starts/tree/readme-badge. Will make a PR after this is merged.

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java (hunk 11 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java]: Done

File: starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java (hunk 10 above)
Comment (jtx1999) [reply starts-core/src/main/java/edu/illinois/starts/helpers/RTSUtil.java]: Done

File: starts-core/pom.xml (hunk 1 above)
Comment (jtx1999) [reply starts-core/pom.xml:40]: I'm getting `[ERROR] /Users/tianxing/Desktop/starts/starts-core/src/main/java/edu/illinois/starts/helpers/ZLCHelper.java:[171,70] method references are not supported in -source 1.7`

File: starts-plugin/pom.xml (hunk 2 above)
Comment (jtx1999) [reply starts-plugin/pom.xml:33]: I'm getting `[ERROR] /Users/tianxing/Desktop/starts/starts-core/src/main/java/edu/illinois/starts/helpers/ZLCHelper.java:[171,70] method references are not supported in -source 1.7`
